protocol/nats_jetstream/v3/protocol.go

```diff
@@ -0,0 +1,310 @@
+		return newValidationError(fieldSendSubject, messageNoSendSubject)
+	}
+
+	defer func() {
+		if err2 := in.Finish(err); err2 != nil {
+			if err == nil {
+				err = err2
+			} else {
+				err = fmt.Errorf("failed to call in.Finish() when error already occurred: %s: %w", err2.Error(), err)
+			}
+		}
+	}()
+
+	if _, err = p.jetStream.StreamNameBySubject(ctx, p.sendSubject); err != nil {
+		return err
+	}
+
+	writer := new(bytes.Buffer)
+	header, err := WriteMsg(ctx, in, writer, transformers...)
+	if err != nil {
+		return err
+	}
+
+	natsMsg := &nats.Msg{
+		Subject: p.sendSubject,
+		Data:    writer.Bytes(),
+		Header:  header,
+	}
+
+	_, err = p.jetStream.PublishMsg(ctx, natsMsg, p.publishOpts...)
+
+	return err
+}
+
+// OpenInbound implements Opener.OpenInbound
+func (p *Protocol) OpenInbound(ctx context.Context) error {
+	p.subMtx.Lock()
+	defer p.subMtx.Unlock()
+
+	var consumeContext jetstream.ConsumeContext
+	var err error
+	if err = p.createJetstreamConsumer(ctx); err != nil {
+		return err
+	}
+	if consumeContext, err = p.jetstreamConsumer.Consume(p.MsgHandler, p.pullConsumeOpts...); err != nil {
+		return err
+	}
+
+	// Wait until external or internal context done
+	select {
+	case <-ctx.Done():
+	case <-p.internalClose:
+	}
+
+	// Finish to consume messages in the queue and close the subscription
+	if consumeContext != nil {
+		consumeContext.Drain()
+	}
+	return nil
+}
+
+// Receive implements Receiver.Receive.
+func (p *Protocol) Receive(ctx context.Context) (binding.Message, error) {
+	select {
+	case msgErr, ok := <-p.incoming:
+		if !ok {
+			return nil, io.EOF
+		}
+		return msgErr.msg, nil
+	case <-ctx.Done():
+		return nil, io.EOF
+	}
+}
+
+// MsgHandler implements nats.MsgHandler and publishes messages onto our internal incoming channel to be delivered
+// via r.Receive(ctx)
+func (p *Protocol) MsgHandler(msg jetstream.Msg) {
+	p.incoming <- msgErr{msg: NewMessage(msg)}
+}
+
+// Close must be called after use to releases internal resources and close the underlying connection
+func (p *Protocol) Close(ctx context.Context) error {
+	// Before closing, let's be sure OpenInbound completes
+	// We send a signal to close and then we lock on subMtx in order
+	// to wait OpenInbound to finish draining the queue
+	p.internalClose <- struct{}{}
+	p.subMtx.Lock()
+	defer p.subMtx.Unlock()
+
+	if p.connOwned && p.conn != nil {
+		p.conn.Close()
+	}
+
+	close(p.internalClose)
+
+	return nil
+}
+
+// applyOptions at the protocol layer should run before the sender and receiver are created.
+// This allows the protocol to create a nats connection that can be shared for both the sender and receiver.
+func (p *Protocol) applyOptions(opts ...ProtocolOption) error {
+	for _, fn := range opts {
+		if err := fn(p); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// validateOptions runs after all options have been applied and makes sure needed options were set correctly.
+func (p *Protocol) validateOptions() error {
```

Comment:
nit: see comment in `options.go` where I suggested to move whole validation logic into a separate file (same package).

Reply:
moved